### PR TITLE
Feature/remaining hooks

### PR DESCRIPTION
**Pull request by patrickbenjamin1**
## What's new?

- Add useInfinitePaging hook - now uses a reducer for its underlying state, and has extensive documentation
- Add useMatchMedia hook
- Add useDidMountEffect and useWillUnmountEffect
- add scroll to validation errors option

## Ticket number(s) in JIRA (if internal)

ARM-170, ARM-117, ARM-127, ARM-125, 

[board](https://rocketmakers.atlassian.net/jira/software/projects/HM/boards/172)

## Checklist

- [x] are your changes in Storybook?
- [ ] are any breaking changes documented in `docs/migrating_from_oldstrong.md`?
- [x] does _everything_ have jsdoc?
- [x] is everything exported from index.ts?

### CODE REVIEW

File: module/src/hooks/useInfinitePaging.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+      try {
+        const response = await fetch(fetchPageToken);
+
+        if (!response) {
+          throw new Error('No response returned from fetch in useInfinitePaging');
+        }
+
+        const noReturnedItems = !response.data || response.data.length === 0;
+        const responseSmallerThanPageSize = settings.pageSize && response.data.length < settings.pageSize;
```
Comment (jlews) [inline module/src/hooks/useInfinitePaging.ts:147]: The above line suggests that `response.data` might be falsy, so `response.data.length` seems risky here.

Suggestion:
```
const responseSmallerThanPageSize = !noReturnedItems && settings.pageSize && response.data.length < settings.pageSize;

```

File: module/src/hooks/useInfinitePaging.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    if (!isFetching && !state.hasFinished) {
+      await fetcher(state.nextPageToken!);
+    }
+  }, [fetcher, isFetching, state.hasFinished, state.nextPageToken]);
+
+  /** reload the state  */
+  const reload = React.useCallback(async () => {
+    dispatch({ type: 'reset', getKey, initialItems: settings.initialItems || [] });
+    await fetcher(settings.firstPageToken);
+  }, [fetcher, dispatch, getKey, settings.initialItems]);
```
Comment (jlews) [inline module/src/hooks/useInfinitePaging.ts:179]: missing dependency: `settings.firstPageToken`

File: module/src/hooks/useWillUnmountEffect.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+/** A useEffect which only runs on the cleanup of the last effect */
+export function useWillUnMountEffect(callback: () => void) {
+  React.useEffect(() => callback, []);
+}
+
+/** A useLayoutEffect which only runs on the cleanup of the last effect */
+export function useWillUnMountLayoutEffect(callback: () => void) {
+  React.useLayoutEffect(() => callback, []);
+}
```
Comment (jlews) [inline module/src/hooks/useWillUnmountEffect.ts:11]: For the same reason as earlier, I would be in favour of this for completeness:
```
/** A useEffect which only runs on the cleanup of the last effect */
export function useWillUnMountEffect(callback: ReturnType<React.EffectCallback>) {
  return React.useEffect(() => callback, []);
}

/** A useLayoutEffect which only runs on the cleanup of the last effect */
export function useWillUnMountLayoutEffect(callback: ReturnType<React.EffectCallback>) {
  return React.useLayoutEffect(() => callback, []);
}
```

**Review by jlews — CHANGES_REQUESTED** [review]

File: module/src/hooks/useDidMountEffect.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+/** A useEffect which will only run in the first effect */
+export function useDidMountEffect(callback: () => void) {
+  React.useEffect(callback, []);
+}
+
+/** A useLayoutEffect which will only run in the first effect */
+export function useDidMountLayoutEffect(callback: () => void) {
+  React.useLayoutEffect(callback, []);
+}
```
[the remark this answers is not in the document]
Comment (patrickbenjamin1) [reply module/src/hooks/useDidMountEffect.ts:11]: I didn't want to use the EffectCallback type cus if you return another function from it it won't just be a did mount effect any more... maybe that's needless restrictiveness but it felt right to make it exclusively able to run on the first effect after mount and not do anything else, given that the hooks are already kind of just restrictions on useEffect.

Take the point about returning the result though

File: module/src/hooks/useWillUnmountEffect.ts (hunk 3 above)
Comment (patrickbenjamin1) [reply module/src/hooks/useWillUnmountEffect.ts:11]: That's groovy

File: module/src/hooks/useDidMountEffect.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jlews) [reply module/src/hooks/useDidMountEffect.ts:11]: Personally I would have expected to be able to return a destructor from `useDidMountEffect` if I wanted to 🤷  but I don't care hugely. It does seem a shame not to though, it just means if you want to add a destructor to an existing `useDidMountEffect` later, you'll get what might seem like a weird typing error before you remember that you have to replace it with a real empty dependency effect for no reason.

**Pull request merged**